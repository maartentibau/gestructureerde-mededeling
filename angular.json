{
  "$schema": "./node_modules/@angular/cli/lib/config/schema.json",
  "version": 1,
  "newProjectRoot": "projects",
  "projects": {
    "gestructureerde-mededeling": {
      "projectType": "application",
      "schematics": {
        "@schematics/angular:component": {
          "style": "scss"
        }
      },
      "root": "",
      "sourceRoot": "src",
      "prefix": "app",
      "architect": {
        "build": {
          "builder": "@angular-devkit/build-angular:application",
          "options": {
            "allowedCommonJsDependencies": ["clipboard"],
            "outputPath": "dist/gestructureerde-mededeling",
            "index": "src/index.html",
            "browser": "src/main.ts",
            "tsConfig": "tsconfig.app.json",
            "inlineStyleLanguage": "scss",
            "assets": ["./src/favicon.ico", "./src/assets", "./src/_redirects"],
            "styles": [
              "normalize.css/normalize.css",
              "@angular/material/prebuilt-themes/azure-blue.css",
              "./src/styles.scss"
            ],
            "scripts": []
          },
          "configurations": {
            "production": {
              "budgets": [
                {
                  "type": "initial",
                  "maximumWarning": "500kB",
                  "maximumError": "1MB"
                },
                {
                  "type": "anyComponentStyle",
                  "maximumWarning": "2kB",
                  "maximumError": "4kB"
                }
              ],
              "outputHashing": "all"
            },
            "development": {
              "optimization": false,
              "extractLicenses": false,
              "sourceMap": true
            }
          },
          "defaultConfiguration": "production"
        },
        "serve": {
          "builder": "@angular-devkit/build-angular:dev-server",
          "configurations": {
            "production": {
              "buildTarget": "gestructureerde-mededeling:build:production"
            },
            "development": {
              "buildTarget": "gestructureerde-mededeling:build:development"
            }
          },
          "defaultConfiguration": "development"
        },
        "extract-i18n": {
          "builder": "@angular-devkit/build-angular:extract-i18n"
        },
        "test": {
          "builder": "@angular-devkit/build-angular:jest",
          "options": {
            "tsConfig": "tsconfig.spec.json"
          }
        }
      }
    }
  }
}
