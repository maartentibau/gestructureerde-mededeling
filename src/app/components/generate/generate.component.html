<ogm-number [ogm]="ogm?.numberFormat"></ogm-number>

<ogm-controls
  [ogm]="ogm"
  [refresh]="true"
  [copyNumber]="true"
  [copyOgm]="true"
  (refreshClick)="refreshClickHandler()"
  (copyNumberClick)="copyNumberClickHandler($event)"
  (copyOgmClick)="copyOgmClickHandler($event)"
></ogm-controls>
