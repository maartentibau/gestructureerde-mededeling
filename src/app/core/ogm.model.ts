export interface OgmData {
  number: string;
  numberFormat: string;
}
