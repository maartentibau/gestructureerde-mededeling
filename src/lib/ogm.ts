export function ogm(): string {
  return 'ogm';
}
