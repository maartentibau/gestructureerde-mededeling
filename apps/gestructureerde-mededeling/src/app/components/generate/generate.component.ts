import { ChangeDetectionStrategy, Component, OnInit } from '@angular/core';
import { MatSnackBar } from '@angular/material/snack-bar';
import { Title } from '@angular/platform-browser';
import { DEFAULT_TITLE } from '../../core/core.constants';
import { OgmService } from '../../core/services/ogm.service';
import { OgmData } from '../../core/ogm.model';

@Component({
  selector: 'ogm-generate',
  templateUrl: './generate.component.html',
  styleUrls: ['./generate.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class GenerateComponent implements OnInit {
  ogm: OgmData | null | undefined;

  constructor(private ogmService: OgmService, private snackBar: MatSnackBar, private title: Title) {
    this.title.setTitle(`${DEFAULT_TITLE} - Genereer een willekeurig gestructureerde mededeling`);
  }

  ngOnInit() {
    this.refreshClickHandler();
  }

  refreshClickHandler() {
    const ogmNumber = this.ogmService.generate();
    const ogmNumberFormat = this.ogmService.format(ogmNumber);

    this.ogm = { number: ogmNumber, numberFormat: ogmNumberFormat };
  }

  copyNumberClickHandler(message: string) {
    this.showSnackbarMessage(message);
  }

  copyOgmClickHandler(message: string) {
    this.showSnackbarMessage(message);
  }

  showSnackbarMessage(message?: string) {
    if (message) {
      this.snackBar.open(message, '', {
        panelClass: 'center-clipboard-text',
        verticalPosition: 'top',
        duration: 750,
      });
    }
  }
}
