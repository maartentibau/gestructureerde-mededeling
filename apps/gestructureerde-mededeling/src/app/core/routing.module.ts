import { NgModule } from '@angular/core';
import { RouterModule, Routes } from '@angular/router';
import { CreateComponent } from '../components/create/create.component';
import { GenerateComponent } from '../components/generate/generate.component';
import { ValidateComponent } from '../components/validate/validate.component';

const routes: Routes = [
  { path: 'create', component: CreateComponent },
  { path: 'generate', component: GenerateComponent },
  { path: 'validate', component: ValidateComponent },
  { path: '', redirectTo: '/generate', pathMatch: 'full' },
  { path: '**', component: GenerateComponent },
];

@NgModule({
  imports: [RouterModule.forRoot(routes, { relativeLinkResolution: 'legacy' })],
  exports: [RouterModule],
})
export class RoutingModule {}
