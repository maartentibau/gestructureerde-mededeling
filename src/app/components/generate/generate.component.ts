import { ChangeDetectionStrategy, Component, inject, OnInit, signal } from '@angular/core';
import { MatSnackBar, MatSnackBarModule } from '@angular/material/snack-bar';
import { Title } from '@angular/platform-browser';
import { ControlsComponent } from '../../core/components/controls/controls.component';
import { NumberComponent } from '../../core/components/number/number.component';
import { DEFAULT_TITLE } from '../../core/core.constants';
import { OgmData } from '../../core/ogm.model';
import { OgmService } from '../../core/services/ogm.service';

@Component({
  standalone: true,
  imports: [NumberComponent, ControlsComponent, MatSnackBarModule],
  selector: 'ogm-generate',
  templateUrl: './generate.component.html',
  styleUrls: ['./generate.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class GenerateComponent implements OnInit {
  #ogmService: OgmService = inject(OgmService);
  #snackBar: MatSnackBar = inject(MatSnackBar);
  #title: Title = inject(Title);

  readonly ogm = signal<OgmData | null>(null);

  constructor() {
    this.#title.setTitle(`${DEFAULT_TITLE} - Genereer een willekeurig gestructureerde mededeling`);
  }

  ngOnInit() {
    this.refreshClickHandler();
  }

  refreshClickHandler() {
    const ogmNumber = this.#ogmService.generate();
    const ogmNumberFormat = this.#ogmService.format(ogmNumber);

    this.ogm.set({ number: ogmNumber, numberFormat: ogmNumberFormat });
  }

  copyNumberClickHandler(message: string) {
    this.showSnackbarMessage(message);
  }

  copyOgmClickHandler(message: string) {
    this.showSnackbarMessage(message);
  }

  showSnackbarMessage(message?: string) {
    if (message) {
      this.#snackBar.open(message, '', {
        panelClass: 'center-clipboard-text',
        verticalPosition: 'top',
        duration: 750,
      });
    }
  }
}
