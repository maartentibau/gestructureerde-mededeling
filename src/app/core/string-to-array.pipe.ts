import { Pipe, PipeTransform } from '@angular/core';

@Pipe({
  standalone: true,
  name: 'stringToArray',
})
export class StringToArrayPipe implements PipeTransform {
  transform(value: number | string | null | undefined): string[] {
    if ((value && typeof value === 'string') || typeof value === 'number') {
      return Array.from(String(value));
    }

    return [];
  }
}
