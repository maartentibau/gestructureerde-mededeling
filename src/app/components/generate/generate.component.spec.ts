import { AsyncPipe, JsonPipe } from '@angular/common';
import { Component, input } from '@angular/core';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { MatSnackBar, MatSnackBarConfig } from '@angular/material/snack-bar';
import { Title } from '@angular/platform-browser';
import { render } from '@testing-library/angular';
import { DEFAULT_TITLE } from '../../core/core.constants';
import { OgmData } from '../../core/ogm.model';
import { OgmService } from '../../core/services/ogm.service';

import { GenerateComponent } from './generate.component';

@Component({
  standalone: true,
  imports: [JsonPipe],
  selector: 'ogm-number',
  template: `
    <div>{{ ogm | json }}</div>
    <div>{{ isValid | json }}</div>
  `,
})
class MockNumberComponent {
  ogm = input<string | null>(null);
  isValid = input<boolean | null>(null);
}

@Component({
  standalone: true,
  imports: [JsonPipe],
  selector: 'ogm-controls',
  template: `
    <div>{{ refresh | json }}</div>
    <div>{{ copyNumber | json }}</div>
    <div>{{ copyOgm | json }}</div>
    <div>{{ ogm | json }}</div>
  `,
})
class MockControlsComponent {
  refresh = input<boolean>(false);
  copyNumber = input<boolean>(false);
  copyOgm = input<boolean>(false);
  ogm = input<OgmData | null>(null);
}

describe('GenerateComponent', () => {
  let component: GenerateComponent;
  let fixture: ComponentFixture<GenerateComponent>;
  let snackBar: MatSnackBar;
  let ogmService: OgmService;
  let titleService: Title;

  const number: string = '123456789012';
  const numberFormat: string = '+++123/4567/89012+++';

  beforeEach(async () => {
    ({ component, fixture, snackBar, ogmService, titleService } = await setup());
  });

  beforeEach(() => {
    jest.spyOn(ogmService, 'generate').mockReturnValueOnce(number);
    jest.spyOn(ogmService, 'format').mockReturnValueOnce(numberFormat);
    jest.spyOn(snackBar, 'open');
  });

  afterEach(() => {
    jest.resetAllMocks();
    jest.restoreAllMocks();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  describe('rendering component', () => {
    it('should match snapshot', () => {
      // act
      fixture.detectChanges();

      // check
      expect(fixture).toMatchSnapshot();
    });
  });

  describe('constructor', () => {
    it('set the correct HTML title', () => {
      // prepare
      const title: string = `${DEFAULT_TITLE} - Genereer een willekeurig gestructureerde mededeling`;
      jest.spyOn(titleService, 'setTitle');

      // act
      TestBed.createComponent(GenerateComponent);

      // check
      expect(titleService.setTitle).toHaveBeenCalledWith(title);
    });
  });

  describe('ngOnInit', () => {
    it('should call refresh method', () => {
      // prepare
      jest.spyOn(component, 'refreshClickHandler');

      // act
      component.ngOnInit();

      // check
      expect(component.refreshClickHandler).toHaveBeenCalled();
    });
  });

  describe('refreshClickHandler', () => {
    it('should refresh the current OGM number', () => {
      // prepare
      const expectedResult = {
        number,
        numberFormat,
      };

      // act
      component.refreshClickHandler();

      // check
      expect(ogmService.generate).toHaveBeenCalled();
      expect(ogmService.format).toHaveBeenCalled();
      expect(component.ogm()).toStrictEqual(expectedResult);
    });
  });

  describe('Copy number or OGM', () => {
    beforeEach(() => {
      jest.spyOn(component, 'showSnackbarMessage');
    });

    describe('copyNumberClickHandler', () => {
      it('should the correct copy message when called', () => {
        // prepare
        const message: string = 'Copy number!';

        // act
        component.copyNumberClickHandler(message);

        // check
        expect(component.showSnackbarMessage).toHaveBeenCalledWith(message);
      });
    });

    describe('copyOgmClickHandler', () => {
      it('should the correct copy message when called', () => {
        // prepare
        const message: string = 'Copy OGM!';

        // act
        component.copyOgmClickHandler(message);

        // check
        expect(component.showSnackbarMessage).toHaveBeenCalledWith(message);
      });
    });
  });

  describe('showSnackbarMessage', () => {
    const snackBarConfig: MatSnackBarConfig = {
      panelClass: 'center-clipboard-text',
      verticalPosition: 'top',
      duration: 750,
    };

    it('should call open when a message is provided', () => {
      // prepare
      const message: string = 'Hello World!';

      // act
      component.showSnackbarMessage(message);

      // check
      expect(snackBar.open).toHaveBeenCalledWith(message, '', snackBarConfig);
    });

    it('should NOT call open when no message is provided', () => {
      // prepare
      const message: string = '';

      // act
      component.showSnackbarMessage(message);

      // check
      expect(snackBar.open).not.toHaveBeenCalled();
    });
  });
});

const setup = async () => {
  const renderResult = await render(GenerateComponent, {
    componentImports: [MockNumberComponent, MockControlsComponent, AsyncPipe],
    providers: [{ provide: MatSnackBar, useValue: { open: jest.fn() } }],
  });

  const snackBar = TestBed.inject(MatSnackBar);
  const ogmService = TestBed.inject(OgmService);
  const titleService = TestBed.inject(Title);

  return {
    ...renderResult,
    component: renderResult.fixture.componentInstance,
    snackBar,
    ogmService,
    titleService,
  };
};
