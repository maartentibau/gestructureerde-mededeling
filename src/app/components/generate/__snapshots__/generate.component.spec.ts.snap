// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenerateComponent rendering component should match snapshot 1`] = `
<ogm-generate
  ogm={[Function Function]}
>
  <ogm-number>
    <div>
    </div>
    <div>
    </div>
  </ogm-number><ogm-controls>
    <div>
    </div>
    <div>
    </div>
    <div>
    </div>
    <div>
    </div>
  </ogm-controls>
</ogm-generate>
`;
